feat(checkins): Participant Check-in Dashboard

harshtandiya opened the pull request:
## What type of PR is this? (check all applicable)
- [x] 🍕Feature

## Description
<!-- Briefly describe the changes introduced by this pull request -->
- Created a dashboard page, which can be used by the chapter members to check-in any participant.

It consists of a list of attendees with details such as "Name", "Ticket ID", "Bought Tshirt" (checkbox which tells whether the person has bought a tshirt along with their ticket, and "Tshirt Provided" (a checkbox which tells whether the tshirt was delivered to the person).

There is a `Checkin Status` column which tells the times (multiple) when a person has been checked in.

There is also a column called `Actions` which has CTAs "Manage" and "Check-In", which open respective dialog boxes as shown in the gif below.

Tshirts can also be marked as "assigned" at the time of checkin or at any later time.

Currently, the check-in system is only feasible for paid events, but we can plan it to move for unpaid events too.

The list can be searched by `Full Name`, `Email` and `Ticket ID` of the attendee.

This dashboard can only be accessed by the Chapter members of the event.

- Added a new doctype `Event Check Ins`. 
This is done because multi-day events require multi-day check-ins. Hence, rather than having a date-time field in the ticket doctype itself, it is more feasible to go with a child table where multiple datetimes can be filled for the same ticket.

- `tshirt_assigned` field is added to the Ticket doctype. If the ticket had a tshirt addon to it, this field marks whether the tshirt was assigned to teh respective person.

- Added a couple of tests for the checkin methods.

## Related Issues & Docs
<!--
For pull requests that relate or close an issue, please include them
below.  We like to follow [Github's guidance on linking issues to pull requests](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

For example having the text: "closes #1234" would connect the current pull
request to issue 1234.  And when we merge the pull request, Github will
automatically close the issue.
-->
closes #596 

## Screenshots/GIFs/Screen Recordings (if applicable)
<!-- Visually demonstrate the changes, if applicable -->
[Screencast from 2024-09-05 15-50-44.webm](https://github.com/user-attachments/assets/c6c7e029-e585-4366-a6ac-84215172428c)

harshtandiya committed:
feat: `Event Check In` child doctype

harshtandiya committed:
feat: Add DoubleChecks icon component

harshtandiya committed:
feat(tickets): add `check_ins` table field to Ticket doctype

harshtandiya committed:
chore(tickets): Refactor ticket authorization logic

harshtandiya committed:
feat(checkin): Add Event Checkin dashboard page

harshtandiya committed:
feat(checkin): Add search functionality to Event Checkin dashboard page

harshtandiya committed:
feat(ticket): Add tshirt_assigned field to Ticket doctype

harshtandiya committed:
chore(checkin): show confirmation dialog before attendee checkin

harshtandiya committed:
feat(checkin): Add CheckinConfirmationDialog component for attendee check-in confirmation

harshtandiya committed:
chore(checkin): Add CheckinManageDialog component for attendee check-in management

harshtandiya committed:
feat(checkin): Add CheckinAttendeeList component for displaying and managing attendee check-ins

harshtandiya committed:
feat(checkin): Refactor EventCheckins.vue component and related files

The commit refactors the EventCheckins.vue component and related files to improve the code structure and organization. It introduces the CheckinAttendeeList component for displaying and managing attendee check-ins, the CheckinManageDialog component for attendee check-in management, and the CheckinConfirmationDialog component for attendee check-in confirmation. It also adds the tshirt_assigned field to the Ticket doctype and implements search functionality in the Event Checkin dashboard page.

harshtandiya committed:
chore(checkin): set `assignTshirt` to false on succesful checkin

harshtandiya committed:
feat(checkin): Add tier information to CheckinConfirmationDialog and CheckinManageDialog components

rahulporuri commented:
general comment - it's time to up the game with PR descriptions. After describing the problem that is being solved in this PR and a description of the specific solution implemented, it will be good to also highlight the technical details e.g. new doctypes that are added, existing doctypes that are changed, changes in roles/permissions to existing doctypes, etc.

rahulporuri commented on fossunited/fossunited/doctype/event_check_in/event_check_in.py:
```diff
@@ -0,0 +1,22 @@
+    if TYPE_CHECKING:
+        from frappe.types import DF
+
+        check_in_time: DF.Datetime | None
```
@harshtandiya  it's not clear to me why we need an entirely new doctype instead of just adding `check_in_time` attribute to the existing `FOSSEventTicket` doctype.

Is the usecase is to enable multiple "check-ins" e.g. for multi-day events? Even in that case, `check_in_times` could be the attribute on the `FOSSEventTicket` doctype which is a list of datetimes or none and everytime check-in happens, the list will be appended to with the new check-in time.

A standalone doctype feels like overkill, maybe because I don't have sufficient context

harshtandiya replied:
The usecase is indeed for multiple check in times for multi-day events.

For a field that is a list of datetimes, i am not sure if frappe would provide a fieldtype which will let us store a "list" of things. The thing that comes the closest is indeed a child doc/table which i did here.

Other than that there's JSON field type, which i feel will be messy in this use case.

Also, we can reuse this doctype if we decide to bring in check-ins for RSVPs too

rahulporuri replied:
> For a field that is a list of datetimes, i am not sure if frappe would provide a fieldtype which will let us store a "list" of things. The thing that comes the closest is indeed a child doc/table which i did here.

okay then.

> Also, we can reuse this doctype if we decide to bring in check-ins for RSVPs too

The only reason we need to do check-ins is to provide meal coupons. Other than that, I don't see a strong reason why we should do them for conferences too. So I highly doubt we want to do it for meetup RSVPs. Also because the meetups will likely not have time to actually take "attendance"

harshtandiya replied:
Makes sense

rahulporuri commented on fossunited/ticketing/doctype/foss_event_ticket/foss_event_ticket.py:
```diff
@@ -35,6 +35,7 @@ class FOSSEventTicket(Document):
         organization: DF.Data | None
         razorpay_payment: DF.Link | None
         tier: DF.Data | None
+        tshirt_assigned: DF.Check
```
By assigned, do you mean distributed or disbursed or delivered? If a person purchases a t-shirt, I would say that a t-shirt is assigned to the ticket. And separately, it is "delivered" to the participant, whether that delivery happens in-person or otherwise.

harshtandiya replied:
Ah yes. I can rename it to "tshirt_delivered" if that would make it more clear?

rahulporuri replied:
yes please. i think it helps. others can chime in and say no

rahulporuri commented on fossunited/ticketing/doctype/foss_event_ticket/foss_event_ticket.json:
```diff
@@ -157,6 +176,12 @@
    "share": 1,
    "write": 1
+  },
+  {
+   "export": 1,
+   "read": 1,
+   "role": "Chapter Lead",
+   "select": 1
```
Is this necessary to ensure that check-ins can be setup? Not sure why this new permission is needed

harshtandiya replied:
Yeah, this is unnecessary now that i see it. Removing this permission

rahulporuri commented on fossunited/api/checkins.py:
```diff
@@ -0,0 +1,100 @@
+
+
+@frappe.whitelist()
+def undo_attendee_checkin(event_id: str, attendee: dict, user: str = frappe.session.user):
```
what is the usecase here? user mistake?

harshtandiya replied:
Yes, if accidentally the wrong attendee is checked in, this can help us undo it through the check-in dashboard

rahulporuri replied:
makes sense. just confirming.

harshtandiya committed:
test(ticket): add `test` for checkin method

harshtandiya committed:
test(ticket): add `test_checkin_as_non_chapter_member` test

This method tests that if any user other than chapter member is trying to check in the user, then an error is thrown

rahulporuri commented on fossunited/ticketing/doctype/foss_event_ticket/test_foss_event_ticket.py:
```diff
@@ -1,8 +1,88 @@
+
+        # Then the check-in data should be saved
+        ticket.reload()
+        self.assertTrue(ticket.check_ins)
```
can we add a few more lines here to do multiple check-ins?

harshtandiya replied:
done

harshtandiya committed:
chore(ticket): remove unwanted permission for `Chapter Member`

harshtandiya committed:
refactor(ticket): refactor `tshirt_assigned` to `tshirt_delivered`

harshtandiya committed:
refactor(checkin): fix typo`checkin_time` -> `check_in_time`

harshtandiya committed:
chore(checkin): add error toast on failed attendee check-in

harshtandiya committed:
chore(checkin): handle already checked in attendee

harshtandiya committed:
test(ticket): add validation for duplicate attendee check-in on the same day

harshtandiya committed:
test(tickets): Add Chapter Team Member and Chapter Lead roles if they don't exist
